api/src/routes/challenge.ts

```diff
@@ -1,24 +1,166 @@
   _options,
   done
 ) => {
+  const challenges = getChallenges();
+
   // eslint-disable-next-line @typescript-eslint/unbound-method
   fastify.addHook('onRequest', fastify.csrfProtection);
   fastify.addHook('onRequest', fastify.authenticateSession);
 
+  fastify.post(
+    '/coderoad-challenge-completed',
+    {
+      schema: schemas.coderoadChallengeCompleted,
+      attachValidation: true
+    },
+    async (req, reply) => {
+      let userToken;
+
+      const { 'coderoad-user-token': encodedUserToken } = req.headers;
+      const { tutorialId } = req.body;
+
+      if (!tutorialId) {
+        void reply.code(400);
+        return {
+          type: 'error',
+          msg: `'tutorialId' not found in request body`
+        } as const;
+      }
+
+      if (!encodedUserToken) {
+        void reply.code(400);
+        return {
+          type: 'error',
+          msg: `'coderoad-user-token' not found in request headers`
+        } as const;
+      }
+
+      try {
+        if (typeof encodedUserToken === 'string' && jwtSecret) {
+          const payload = jwt.verify(encodedUserToken, jwtSecret) as JwtPayload;
+          userToken = payload.userToken;
+        }
+      } catch {
+        void reply.code(400);
+        return { type: 'error', msg: `invalid user token` } as const;
+      }
+
+      const tutorialRepo = tutorialId?.split(':')[0];
+      const tutorialOrg = tutorialRepo?.split('/')?.[0];
+
+      if (tutorialOrg !== 'freeCodeCamp') {
+        void reply.code(400);
+        return {
+          type: 'error',
+          msg: `Tutorial not hosted on freeCodeCamp GitHub account`
+        } as const;
+      }
+
+      const codeRoadChallenges = challenges.filter(
+        ({ challengeType }) => challengeType === 12 || challengeType === 13
+      );
+
+      const challenge = codeRoadChallenges.find(challenge => {
+        return challenge.url?.endsWith(tutorialRepo ?? '');
+      });
+
+      if (!challenge) {
+        void reply.code(400);
+        return { type: 'error', msg: 'Tutorial name is not valid' } as const;
+      }
+
+      const { id: challengeId, challengeType } = challenge;
+      try {
+        const tokenInfo = await fastify.prisma.userToken.findUnique({
+          where: { id: userToken }
+        });
+
+        if (!tokenInfo)
+          return { type: 'error', msg: 'User token not found' } as const;
+
+        const { userId } = tokenInfo;
+
+        const user = await fastify.prisma.user.findFirstOrThrow({
+          where: { id: userId }
+        });
+
+        if (!user)
+          return {
+            type: 'error',
+            msg: 'User for user token not found'
+          } as const;
+
+        const completedDate = Date.now();
+        const { completedChallenges = [], partiallyCompletedChallenges = [] } =
+          user;
+
+        const isCompleted = completedChallenges.some(
+          challenge => challenge.id === challengeId
+        );
+
+        if (challengeType === 13 && !isCompleted) {
```

Comment:
should this wait until challengeTypes is in config folder, https://github.com/freeCodeCamp/freeCodeCamp/pull/51027